### PR TITLE
Removing .git dirs requires a bit more effort on windows.

### PR DESCRIPTION
**Pull request by garyvdm**
shutils.rmtree on .git dirs fails on windows due to the read only attribute being set, so we need to do some additional work to unset the attribute.

### CODE REVIEW

**Comment by jelmer** [comment]
This is specific to the compat tests I think? Dulwich itself doesn't create read-only directories. Please move it to dulwich/tests/compat/utils.py

**Comment by jelmer** [comment]
Please use a different name, as this is doing something different than shutil.rmtree.

**Comment by garyvdm** [comment]
> This is specific to the compat tests I think? Dulwich itself doesn't create read-only directories. Please move it to dulwich/tests/compat/utils.py

Sorry, I was painting with a very broad brush. I've removed a bunch of unnecessary changes. It is still need in some of the non compat tests, e.g. `dulwich.tests.test_porcelain.PushTests.test_simple`, so I've moved the new method to `dulwich/tests/utils.py` rather than `dulwich/tests/compat/utils.py`. 

> Please use a different name, as this is doing something different than shutil.rmtree.

Done.

**Comment by garyvdm** [comment]
Sorry. Made a rebase error, and included other stuff. Fixed now.

File: dulwich/tests/test_porcelain.py [hunk 1]
```diff
@@ -47,7 +48,7 @@ class PorcelainTestCase(TestCase):
     def setUp(self):
         super(TestCase, self).setUp()
         repo_dir = tempfile.mkdtemp()
-        self.addCleanup(shutil.rmtree, repo_dir)
+        self.addCleanup(rmtree_ro, repo_dir)
```
Comment (jelmer) [inline dulwich/tests/test_porcelain.py]: why is this necessary? dulwich doesn't mark any files readonly AFAIK.

Comment (garyvdm) [inline dulwich/tests/test_porcelain.py]: It's needed for `test_porcelain.PushTests.test_simple` which calls a `SubprocessGitClient` when it calls `push`. 

We can change the `default_local_git_client_cls` to `LocalGitClient` for this test, but I don't think there is any harm in using `rmtree_ro` here.

**Comment by garyvdm** [comment]
Opps, I seem to have confused travis, and it's not testing the right sha. Here is a test result of the correct sha on my account: https://travis-ci.org/garyvdm/dulwich/builds/60458018

**Comment by jelmer** [comment]
Thanks for the updates. I think we should update the default push to use the LocalGitClient by default as well. The noncampat testsuite should still work and behave the same independent of whether C git is installed.

On 28 April 2015 11:54:17 pm BST, Gary van der Merwe <notifications@github.com> wrote:
>> This is specific to the compat tests I think? Dulwich itself doesn't
>create read-only directories. Please move it to
>dulwich/tests/compat/utils.py
>
>Sorry, I was painting with a very broad brush. I've removed a bunch of
>unnecessary changes. It is still need in some of the non compat tests,
>e.g. `dulwich.tests.test_porcelain.PushTests.test_simple`, so I've
>moved the new method to `dulwich/tests/utils.py` rather than
>`dulwich/tests/compat/utils.py`. 
>
>> Please use a different name, as this is doing something different
>than shutil.rmtree.
>
>Done.
>
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/jelmer/dulwich/pull/275#issuecomment-97250478

Jelmer Vernooij <jelmer@jelmer.co.uk> - https://jelmer.co.uk/

**Comment by garyvdm** [comment]
> I think we should update the default push to use the LocalGitClient by default as well.

Done.

`rmtree_ro` is still in `tests/utils.py` and not `tests/compat/utils.py` as `dulwich.tests.compat.test_repository.ObjectStoreTestCase` uses `test.utils.tear_down_repo`.

**Comment by jelmer** [comment]
On Thu, Apr 30, 2015 at 02:59:59AM -0700, Gary van der Merwe wrote:
> > I think we should update the default push to use the LocalGitClient by default as well.
> 
> Done.
Thanks!

> `rmtree_ro` is still in `tests/utils.py` and not `tests/compat/utils.py` as `dulwich.tests.compat.test_repository.ObjectStoreTestCase` uses `test.utils.tear_down_repo`.  
Fair enough. 

I'll merge this evening.

File: dulwich/client.py [hunk 2]
```diff
@@ -704,7 +704,7 @@ def send_pack(self, path, determine_wants, generate_pack_contents,
 
         target = Repo(path)
         old_refs = target.get_refs()
-        new_refs = determine_wants(old_refs)
+        new_refs = determine_wants(dict(old_refs))
```
Comment (jelmer) [inline dulwich/client.py]: why is this necessary?

File: dulwich/tests/test_client.py [hunk 3]
```diff
@@ -408,6 +408,12 @@ def test_http(self):
         self.assertTrue(isinstance(c, HttpGitClient))
         self.assertEqual('/jelmer/dulwich', path)
 
+    def test_local_specified_cls(self):
+        c, path = get_transport_and_path(
+            'foo.bar/baz', local_git_client_cls=SubprocessGitClient)
```
Comment (jelmer) [inline dulwich/tests/test_client.py]: won't this break on systems without C git installed?

File: dulwich/tests/test_porcelain.py [hunk 4]
```diff
@@ -465,12 +465,12 @@ def test_simple(self):
         # Check that the target and source
         r_clone = Repo(clone_path)
 
+        self.assertEqual(r_clone[b'HEAD'].id, self.repo[refs_path].id)
```
Comment (jelmer) [inline dulwich/tests/test_porcelain.py]: why does this ordering need to change?

**Comment by jelmer** [comment]
Thanks, I've merged the first commit - added comments to the second.

File: dulwich/tests/test_client.py (hunk 3 above)
Comment (garyvdm) [inline dulwich/tests/test_client.py]: No. This test passes because we are just asserting on the type of client returned. It never calls any methods on the clients, so the client never calls git.

This commit actually fixes a bunch of tests from `dulwich.tests.test_porcelain` that were failing when git was not installed because the default client was `SubprocessGitClient`, and we were not checking if it was installed with `require_git_version`.

File: dulwich/tests/test_porcelain.py (hunk 4 above)
Comment (garyvdm) [inline dulwich/tests/test_porcelain.py]: The test was failing at one point, and the failure message was a confusing `IndexError` from the `change = list(tree_changes(...))[0]` line because there were no changes.

The change in order is so we first check that the push happened by checking the `HEAD` changed, so that the failure message is more understandable.

**Comment by garyvdm** [comment]
I've rebased the `Switched `default_local_git_client_cls` to `LocalGitClient`.` commit.

**Comment by garyvdm** [comment]
Rebased onto master.

**Comment by garyvdm** [comment]
Bump. There is still a commit from this branch not yet merged.

**Comment by garyvdm** [comment]
Thanks.
